datafusion/core/src/execution/context/mod.rs

```diff
@@ -1329,6 +1329,8 @@ pub struct SessionState {
     table_functions: HashMap<String, Arc<TableFunction>>,
     /// Scalar functions that are registered with the context
     scalar_functions: HashMap<String, Arc<ScalarUDF>>,
+    // Need to implement Clone for this, so probably not a good idea.
+    // scalar_functions_impl: HashMap<String, ScalarFactory>,
```

Comment:
It isn't clear to me if @viirya 's usecase needs the defer udf creation in all cases. If the goal is really to avoid instantiting *all* `ScalarUDF`s I think a better approach would be to be to chang the signature of `FunctionRegistry`

However, we may be able to avoid changing session state at all (and simply have an API to get the function factories in addition to registering all functions with the SessionState)